### PR TITLE
rune/libenclave/skeleton && sgx-tools: Solve the launch token adaptat…

### PR DESCRIPTION
**Pull request by haosanzi**
…ion problem of sgx2 + in-tree

- sgx-tools gen-token command reports an error when running on FLC machines
- Operations related to lanunch token are not supported on FLC machines

Signed-off-by: Shirong Hao shirong@linux.alibaba.com

### CODE REVIEW

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile [hunk 1]
```diff
@@ -9,17 +9,16 @@ ENCL_CFLAGS := -Wall -Werror -static -nostdlib -nostartfiles -fPIC \
 	       -fno-stack-protector -mrdrnd -std=gnu11
 HOST_LDFLAGS := -fPIC -shared -Wl,-Bsymbolic
 
-IS_OOT_DRIVER := $(shell [ ! -e /dev/isgx ])
-IS_SGX_FLC := $(shell lscpu | grep -q sgx_lc)
+IS_OOT_DRIVER ?= $(shell if [ -e /dev/isgx ]; then echo "1"; else echo "0"; fi;)
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:12]: IS_OOT_DRIVER is not used anywhere. Delete it plz.

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile [hunk 2]
```diff
@@ -9,17 +9,16 @@ ENCL_CFLAGS := -Wall -Werror -static -nostdlib -nostartfiles -fPIC \
...
 PRODUCT_ENCLAVE ?=
 
 TEST_CUSTOM_PROGS := $(OUTPUT)/encl.bin $(OUTPUT)/encl.ss $(OUTPUT)/liberpal-skeleton-v1.so $(OUTPUT)/liberpal-skeleton-v2.so $(OUTPUT)/liberpal-skeleton-v3.so $(OUTPUT)/signing_key.pem
 
-ifeq ($(IS_OOT_DRIVER),1)
-    TEST_CUSTOM_PROGS += $(OUTPUT)/encl.token
-else ifeq ($(IS_SGX_FLC),)
+ifeq ($(IS_SGX_FLC),0)
     TEST_CUSTOM_PROGS += $(OUTPUT)/encl.token
+    HOST_CFLAGS += -DDEBUG_ENCLAVE
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:21]: Why enclave must be debug if flc is unavailable?

File: rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c [hunk 3]
```diff
@@ -358,6 +358,7 @@ static bool load_sigstruct(const char *path, void *sigstruct)
 	return true;
 }
 
+#ifdef DEBUG_ENCLAVE
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c:361]: This is not right way. The detection of the necessity of launch token **MUST** be dynamic. You change the flexibility in a static way. Plz remove DEBUG_ENCLAVE.

File: sgx-tools/gen-token.go [hunk 4]
```diff
@@ -32,6 +32,11 @@ For example, generate the token file according to the given signature file:
 		},
 	},
 	Action: func(context *cli.Context) error {
+
+		if intelsgx.IsSGXLaunchControlSupported() {
```
Comment (jiazhang0) [inline sgx-tools/gen-token.go:36]: So you did well here with a runtime detection but why applying a static form in skeleton?

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c [hunk 5]
```diff
@@ -88,3 +88,15 @@ void get_sgx_xfrm_by_cpuid(uint64_t *xfrm)
...
+bool is_launch_control_supported(void)
+{
+	int cpu_info[4] = {0, 0, 0, 0};
+
+	__cpuidex(cpu_info, CPUIID_EXTENDED_FEATURE_FLAGS, 0);
+
+	if ((cpu_info[2] & 0x40000000) != 0)
+		return true;
+
+	return false;
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c:101]: Change to:
```C
return !!(cpu_info[2] & 0x40000000);
```

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile [hunk 6]
```diff
@@ -9,16 +9,13 @@ ENCL_CFLAGS := -Wall -Werror -static -nostdlib -nostartfiles -fPIC \
...
+IS_SGX_FLC ?= $(shell if lscpu | grep -q sgx_lc; then echo "1"; else echo "0"; fi;)
 
 PRODUCT_ENCLAVE ?=
 
 TEST_CUSTOM_PROGS := $(OUTPUT)/encl.bin $(OUTPUT)/encl.ss $(OUTPUT)/liberpal-skeleton-v1.so $(OUTPUT)/liberpal-skeleton-v2.so $(OUTPUT)/liberpal-skeleton-v3.so $(OUTPUT)/signing_key.pem
 
-ifeq ($(IS_OOT_DRIVER),1)
-    TEST_CUSTOM_PROGS += $(OUTPUT)/encl.token
-else ifeq ($(IS_SGX_FLC),)
+ifeq ($(IS_SGX_FLC),0)
```
Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:18]: It's better to add below in Makefile
ifeq ($(IS_SGX_FLC),0)
$(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
          sgx-tools gen-token --signature encl.ss --token $@
endif

File: rune/libenclave/internal/runtime/pal/skeleton/sgx.h [hunk 7]
```diff
@@ -19,6 +19,9 @@ enum sgx_page_flags {
 
 #define	SGX_LEAF	0x12
 
+// CPUID leafs
+#define CPUIID_EXTENDED_FEATURE_FLAGS	0x7
+
```
Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/sgx.h:24]: The format issue.  Please use tab intead of space.

**Review by YangLiang3 — COMMENTED** [review]
1.  In Makefile 
$(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
            sgx-tools gen-token --signature encl.ss --token $

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile (hunk 6 above)
Comment (YangLiang3) [reply rune/libenclave/internal/runtime/pal/skeleton/Makefile:18]: $(OUTPUT)/liberpal-skeleton*.o \
$(OUTPUT)/liberpal-skeleton*.so \
$(OUTPUT)/signing_key.pem \
ifeq ($(IS_SGX_FLC),0)
$(OUTPUT)/encl.token 
endif

ifeq ($(IS_SGX_FLC),0)
        EXTRA_CLEAN += $(OUTPUT)/encl.token
endif

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile [hunk 8]
```diff
@@ -9,16 +9,13 @@ ENCL_CFLAGS := -Wall -Werror -static -nostdlib -nostartfiles -fPIC \
...
     TEST_CUSTOM_PROGS += $(OUTPUT)/encl.token
 endif
```
Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:21]: it's better to add  as below in Makefile
ifeq ($(IS_SGX_FLC),0)
$(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
        sgx-tools gen-token --signature encl.ss --token $@
endif

ifeq ($(IS_SGX_FLC),0)
EXTRA_CLEAN += $(OUTPUT)/encl.token
endif

Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:21]: it's better to add as below in Makefile
ifeq ($(IS_SGX_FLC),0)
$(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
sgx-tools gen-token --signature encl.ss --token $@
endif

ifeq ($(IS_SGX_FLC),0)
EXTRA_CLEAN += $(OUTPUT)/encl.token
endif

File: sgx-tools/gen-token.go [hunk 9]
```diff
@@ -32,6 +32,11 @@ For example, generate the token file according to the given signature file:
...
+			return fmt.Errorf("It is not suitable to run the sgx-tools gen-token command on flc machines")
```
Comment (jiazhang0) [inline sgx-tools/gen-token.go:37]: Change to:
```
"gen-token command is unable to run without SGX launch control feature"
```

File: rune/libenclave/internal/runtime/pal/skeleton/Makefile [hunk 10]
```diff
@@ -62,16 +59,18 @@ $(OUTPUT)/signing_key.pem:
...
+        PRODUCT_OPT :=
 endif
 
 $(OUTPUT)/encl.ss: $(OUTPUT)/encl.bin $(OUTPUT)/signing_key.pem
 	$(OUTPUT)/sgxsign $(PRODUCT_OPT) signing_key.pem $(OUTPUT)/encl.bin $(OUTPUT)/encl.ss
 
-$(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
-	sgx-tools gen-token --signature encl.ss --token $@
+ifeq ($(IS_SGX_FLC),0)
+        $(OUTPUT)/encl.token: $(OUTPUT)/encl.ss
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/skeleton/Makefile:71]: This indent is not necessary for rule.

**Review by jiazhang0 — APPROVED** [review]

**Pull request merged**